File: 2009/S093.pas
var
  right: array[0..2999] of int64;
  n, m, a, b, i, j: longint;
  max: int64;
begin
  for i := 0 to 2999 do
    right[i] := 0;
  read(n, m);
  for i := 1 to n do
  begin
    read(a, b);
    max := 0;
    for j := m - 1 downto -1 do
    begin
      if (j = -1) then
        break;
      if (right[j] > max) then
      begin
        if (m - j - 1 >= b) then
          break;
        max := right[j];
      end
    end;
    right[j + b] := max + a;
    writeln(max, ' ', j + 1)
  end
end.

File: 2009/S093.cpp
#include <cstdio>
long long right[3000];
int main() {
  int n, m, a, b;
  for (int i = 0; i < 3000; i++) {
    right[i] = 0;
  }
  scanf("%d %d", &n, &m);
  for (int i = 0; i < n; i++) {
    scanf("%d %d", &a, &b);
    long long max = 0;
    int j;
    for (j = m - 1; j >= 0; j--) {
      if (right[j] > max) {
        if (m - j - 1 >= b)  {
          break;
        }
        max = right[j];
      }
    }
    right[j + b] = max + a;
    printf("%lld %d\n", max, j + 1);
  }
  return 0;
}
